R/dsmodel.R

```diff
@@ -335,31 +334,76 @@ dsmodel <- function(fun, title="", display = TRUE) {
-		  #    x=x$x
-		  #  }
-		  #  else if(is.vector(x) && length(x)==2){
-		  #    y=x[[2]]
-		  #    x=x[[1]]
-		  #  }
-		  #  else {
-	  	#    stop("dsmodel: expected input formats for find.period's starting point are two scalars, a vector of length two or a dspoint")
-		  #  }
-		  #}
+		find.period= function(self, a, b, x=NULL, y=NULL, iters=1000, maxPeriod=128, initIters=0, numTries=1, powerOf2=TRUE,
+                          epsilon=sqrt(sqrt(.Machine$double.eps)), crop=FALSE, xlim=NULL, ylim=NULL, aname=NULL, bname=NULL){
+		  dsassert(is.paramrange(self$range), paste(      #paste is the only way to make multi line strings that dont contain newlines.
+		           "to use find.period model's range must be a paramRange. Most likely, ",
+		           "you are getting this error message becuse an object you added to the model ",
+		           "uses find.period."))
 		  if(crop){
-		    dsassert(self$has.xyrange(),"Finding period with crop set to true requires the model's range to be defined.")
+		    hasLims=self$has.xyrange() ||(!is.null(xlim) && !is.null(ylim))
+		    dsassert(hasLims, paste("Finding period with crop set to true either ",
+                 "requires the model's x and y lims to be defined, or ",
+                 "requires x and y lims to be set in the relevant object."))
+		  }
+		  if(is.null(x)) {
+		    if(self$has.xyrange()) { #take a point from near the front of the model's range
+		      rangexlim=self$range$xlim
+		      rangeylim=self$range$ylim
+		      x=(rangexlim[2]-rangexlim[1])/100+rangexlim[1]
+		      y=(rangeylim[2]-rangeylim[1])/100+rangeylim[1]
+		    }
+		    else if(!is.null(xlim) && !is.null(ylim)){ #take a point from near the front of the passed in range
+		      x=(xlim[2]-xlim[1])/100+xlim[1]
+		      y=(ylim[2]-ylim[1])/100+ylim[1]
+		    }
+		    else if(!is.null(self$range$discretize) && self$range$discretize != 0) { #take a point close to (0,0), use discretize for scale
+		      x=self$range$discretize/2
+		      y=self$range$discretize/2
+		    }
+		    else { #take a point close to (0,0)
+		      x=.05
+		      y=.05
+		    }
+		  }
+		  else if(!(!is.null(y) && length(x)==1 && length(y)==1)){ # y isn't a scalar, try to pull x and y values from x.
+		    if(is.dspoint(x)){
+		      y=x$y
+		      x=x$x
+		    }
+		    else if(is.vector(x) && length(x)==2){
+		      y=x[[2]]
+		      x=x[[1]]
+		    }
+		    else {
+	  	    stop("dsmodel: expected input formats for find.period's starting point are two scalars, a vector of length two or a dspoint")
+		    }
 		  }
+		  if(is.null(aname))
+  		  aname=self$range$aname
+		  if(is.null(bname))
+		    bname=self$range$bname
+		  args=list(FUN=self$find.period.internal, MoreArgs=list(x=x, y=y, iters=iters, maxPeriod=maxPeriod,
+		            initIters=initIters, numTries=numTries, powerOf2=powerOf2, epsilon=epsilon,
+		            crop=crop, xlim=xlim, ylim=ylim))
+		  args[[aname]]=a
+		  args[[bname]]=b
+      ret=do.call(what=mapply,args=args)

```

Comment:
Is the only case where we need mapply when x and y are passed in with length > 2? Is that a thing we currently do in the code?